src/clj/puppetlabs/puppetserver/shell_utils.clj

```diff
@@ -0,0 +1,96 @@
+(ns puppetlabs.puppetserver.shell-utils
+  (:require [schema.core :as schema]
+            [clojure.java.io :as io])
+  (:import (com.puppetlabs.puppetserver ProcessWrapper ExecutionStubImpl)
+           (org.apache.commons.exec CommandLine DefaultExecutor PumpStreamHandler)
+           (java.io IOException ByteArrayInputStream ByteArrayOutputStream)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Schemas
+
+(def ExecutionResult
+  "A map that contains the details of the result of executing a command."
+  {:exit-code schema/Int
+   :stderr schema/Str
+   :stdout schema/Str})
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Internal
+
+(schema/defn ^:private ^:always-validate
+  validate-command
+  "Checks the command string to ensure that is is an absolute path, executable
+  and that the file exists. An exception is thrown is any of those are not the
+  case."
+  [command :- schema/Str]
+  (let [command-file (io/as-file command)]
+    (cond
+      (not (.isAbsolute command-file))
+      (throw (IllegalArgumentException.
+              (format "An absolute path is required, but %s is not an absolute path" command)))
+      (not (.exists command-file))
+      (throw (IllegalArgumentException.
+              (format "The referenced command '%s' does not exist" command)))
+      (not (.canExecute command-file))
+      (throw (IllegalArgumentException.
+              (format "The referenced command '%s' is not executable" command))))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public
+
+(schema/defn ^:always-validate
+  execute-command :- ExecutionResult
+  "Execute the specified fully qualified command and any included options and
+  return the exit-code, and the contents of stdout and stderr for the command."
+  ([command :- schema/Str]
+   (execute-command command []))
+  ([command :- schema/Str
+    options :- [schema/Str]]
+   (validate-command command)
+   (try
+     (let [process (ProcessWrapper. (. (Runtime/getRuntime) exec (into-array (concat [command] options))))]
+       {:exit-code (.getExitCode process)
+        :stderr (.getErrorString process)
+        :stdout (.getOutputString process)})
+     (catch IOException e
+       (throw (IllegalStateException. (format "Exception while executing '%s': %s" command (.getMessage e))))))))
+
+(schema/defn ^:always-validate
+execute-command' :- ExecutionResult
```

Comment:
Whoa, didn't notice the `'` at the end of the function name.  

If this is meant to be a non-public or subordinate function that's used by `execute-command`, then there are existing naming patterns like `do-execute-command` or similar that we should probably stick to.

If this is supposed to be an alternate execute-command that deals with streams instead of strings, we should at least mention that in the docstring and probably rename it appropriately too